src/multistream_select/dialer_select.rs

```diff
@@ -269,29 +270,22 @@ impl DialerState {
         protocol: ProtocolName,
         fallback_names: Vec<ProtocolName>,
     ) -> crate::Result<(Self, Vec<u8>)> {
-        // encode `/multistream-select/1.0.0` header
-        let mut bytes = BytesMut::with_capacity(64);
-        let message = Message::Header(HeaderLine::V1);
-        let _ = message.encode(&mut bytes).map_err(|_| Error::InvalidData)?;
-        let mut header = UnsignedVarint::encode(bytes)?;
-
-        // encode proposed protocol
-        let mut proto_bytes = BytesMut::with_capacity(512);
-        let message = Message::Protocol(Protocol::try_from(protocol.as_bytes()).unwrap());
-        let _ = message.encode(&mut proto_bytes).map_err(|_| Error::InvalidData)?;
-        let proto_bytes = UnsignedVarint::encode(proto_bytes)?;
-
-        // TODO: add fallback names
-
-        header.append(&mut proto_bytes.into());
+        let message = encode_multistream_message(
+            std::iter::once(protocol.clone())
+                .chain(fallback_names.clone())
+                .filter_map(|protocol| Protocol::try_from(protocol.as_ref()).ok())
+                .map(|protocol| Message::Protocol(protocol)),
+        )?
```

Comment:
I'll have to check this but this works with smoldot, both for listener and dialer.

Reply:
You're right. It looks like an incorrect way of handling the handshake and litep2p [would reject](https://github.com/altonen/litep2p/blob/d3b1be62904a98b19d03c0bf213658f9f441cf71/src/multistream_select/listener_select.rs#L259) this message as well but smoldot both sends and accepts `Message::Protocols`. I'm kind of tempted to keep it as is since it works but I'll check if smoldot would work with `Message::Protocol`